Issue #8211 - fix ISE from HttpChannelState if failure during process

lachlan-roberts opened the pull request:
If a timeout or stream reset occurs during process, the `HttpChannelState` will be failed, this causes `callback.succeeded()` to throw ISE. This ISE is then caught in `HandlerInvoker` and the callback is failed, and because the callback was failed twice we get the `java.lang.IllegalStateException: already completed` seen in #8211.

This PR fixes this by 
- only failing the callback in `HandlerInvoker` if `HttpChannel` has not been completed.
- Do not throw from `callback.failed()` if the callback is failed with same cause as why the `HttpChannel` was failed.

lachlan-roberts committed:
Issue #8211 - fix ISE from HttpChannelState if failure during process

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -663,7 +663,15 @@ public void run()
             }
             if (failure != null)
-                request._callback.failed(failure);
+            {
+                try (AutoLock ignored = _lock.lock())
+                {
+                    if (request._httpChannel != null && !request._httpChannel._completed)
+                        request._callback.failed(failure);
+                    else if (LOG.isDebugEnabled())
+                        LOG.debug("Process failed", failure);
+                }
+            }
```
Potentially it would be neater to put this logic into a method:
```suggestion
            request._callback.ensureCompleted(failure);
```

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1364,6 +1372,9 @@ public void failed(Throwable failure)
             try (AutoLock ignored = _request._lock.lock())
             {
+                request = _request;
+                if (request._httpChannel != null && request._httpChannel._error != null && TypeUtil.isAssociated(request._httpChannel._error.getCause(), failure))
+                    return;
```
Need a comment here about the fact we are being forgiving of lazy applications that doubly report a failure.
```suggestion
                // Be forgiving of applications that report a failure that is already known by treating as a noop.
                if (request._httpChannel != null && request._httpChannel._error != null && TypeUtil.isAssociated(request._httpChannel._error.getCause(), failure))
                    return;
```

gregw reviewed (approved):
I think this LG, but as I'm kind of a co-author, we should get @sbordet's thoughts as well.

gregw reviewed (dismissed):
oops I approved when I wanted to request some small changes.

lachlan-roberts committed:
Issue #8211 - changes from review

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-8211-channelAlreadyComplete

lachlan-roberts committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-8211-channelAlreadyComplete

lachlan-roberts committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into jetty-12.0.x-8211-channelAlreadyComplete

lachlan-roberts committed:
fixes after merge from jetty-12.0.x

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
try different approach for preventing multiple ChannelCallback completions

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

sbordet commented on jetty-core/jetty-http2/jetty-http2-tests/src/test/java/org/eclipse/jetty/http2/tests/IdleTimeoutTest.java:
```diff
@@ -384,6 +386,77 @@ public boolean onIdleTimeout(Stream stream, Throwable x)
+    @Test
+    public void testForgivingFailedClientEnforcingStreamIdleTimeout() throws Exception
+    {
+        final int idleTimeout = 1000;
```
No need for `final`, here and below.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1317,6 +1319,10 @@ public void succeeded()
             {
+                if (_completed)
+                    return;
+                _completed = true;
```
I would not do this check here.
The call below `lockedOnComplete()` is the one responsible for setting `_completed` to `true`, and throwing if it's already set.
I would: A) make `lockedOnComplete()` return a `boolean`, and if it is `false` return; and B) remove the code in `lockedOnComplete()` that throws, which is now never executed.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -1377,6 +1383,10 @@ public void failed(Throwable failure)
             {
+                if (_completed)
+                    return;
+                _completed = true;
```
Ditto.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -663,8 +663,9 @@ public void run()
+
             if (failure != null)
-                request._callback.failed(failure);
+                _request._callback.ensureCompleted(failure);
```
I am not sure I understand this. With the current modifications, the original code calling `failed()` is already doing the check on `_completed`, so this method does nothing different, and certainly does not "ensure" and it only fails the callback, while the name "ensureCompleted" seems to hint that it either succeeds or fails it.
I would restore the old code.

sbordet committed:
Merged remote-tracking branch 'origin/jetty-12.0.x' into 'jetty-12.0.x-8211-channelAlreadyComplete'.

sbordet committed:
Avoid using another ChannelCallback._completed field when there is already HttpChannelState._completed.
Now all the logic for completion is in lockedOnComplete().

Avoid throwing in ChannelCallback.succeeded() if HttpChannelState._error != null.
This is necessary because HttpChannelState._error may be set asynchronously by
some event such as HTTP/2 reset streams or idle timeouts, but if there is a
thread dispatched to the application the asynchronous event will not fail the
callback, as the failure may be noticed by the application (e.g. via a read() call).

Fixed TrailersTest.testHandlerRequestTrailers() by avoid reading again after
having read the trailers.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
@lachlan-roberts I have pushed changes that align to my comments, so I'll merge this PR once build passes.
